backend/src/halo2/circuit_builder.rs

```diff
@@ -406,31 +443,25 @@ impl<'a, T: FieldElement, F: PrimeField<Repr = [u8; 32]>> Circuit<F> for PowdrCi
                 } else {
                     Some(&new_witness)
                 };
-                for (name, &column) in config.advice.iter() {
-                    // Note that we can't skip this loop if we don't have a witness,
-                    // because the copy_advice() call below also adds copy constraints,
-                    // which are needed to compute the verification key.
-                    let values = witness.as_ref().and_then(|witness| {
-                        witness.iter().find_map(|(witness_name, values)| {
-                            (witness_name == name).then_some(values)
-                        })
-                    });
-                    for i in 0..degree {
-                        let value = values
-                            .map(|values| Value::known(convert_field::<T, F>(values[i])))
-                            .unwrap_or_default();
-
-                        let assigned_cell = region.assign_advice(|| name, column, i, || value)?;
-
-                        // The first row needs to be copied to row <degree>
-                        if i == 0 {
-                            assigned_cell.copy_advice(|| name, &mut region, column, degree)?;
-                        }
-
-                        // Collect public cells, which are later copy-constrained to equal
-                        // a cell in the instance column.
-                        if let Some(&instance_index) = publics.get(&(name.clone(), i)) {
-                            public_cells.push((instance_index, assigned_cell));
+                if let Some(witness) = witness {
+                    for (name, values) in witness.iter() {
+                        let column = *config.advice.get(name).unwrap();
+                        for i in 0..degree {
+                            let value = Value::known(convert_field::<T, F>(values[i]));
+
+                            let assigned_cell =
+                                region.assign_advice(|| name, column, i, || value)?;
+
+                            // The first row needs to be copied to row <degree>
+                            if i == 0 {
+                                region.assign_advice(|| name, column, degree, || value)?;
```

Comment:
Assigning the wrong value here leads to a failure in proof generation, as expected:

```
error: constraint not satisfied

  Cell layout in region 'main':
    | Offset |main.reg_write_X_CNT|__first_step|
    +--------+--------------------+------------+
    |    0   |         x0         |     x2     | <--{ Gate 'enforce_wrapping' applied here
    |  1024  |         x1         |            |

  Constraint 'enforce wrapping (main.reg_write_X_CNT)':
    x2 * (x0 - x1) = 0

  Assigned cell values:
    x0 = 0
    x1 = 1
    x2 = 1
```

Reply:
Nice!